test/tutorials/lambda_theory.jl

```diff
@@ -102,47 +102,55 @@ end
 
 const LambdaAnalysis = Set{Symbol}
 
-getdata(eclass) = isnothing(eclass.data) ? LambdaAnalysis() : eclass.data
+getdata(eclass) = eclass.data
 
 function EGraphs.make(g::EGraph{ExprType,LambdaAnalysis}, n::VecExpr) where {ExprType}
-  v_isexpr(n) || LambdaAnalysis()
+  v_isexpr(n) || return LambdaAnalysis()
   if v_iscall(n)
     h = v_head(n)
     op = get_constant(g, h)
     args = v_children(n)
     eclass = g[args[1]]
-    free = getdata(eclass)
+    free = copy(getdata(eclass))
 
     if op == Variable
       push!(free, get_constant(g, v_head(eclass.nodes[1])))
-
     elseif op == Let
-      v, a, b = args[1:3]
+      v, a, b = args[1:3] # v=a in b
+      vclass = g[v]
+      vsy = get_constant(g, v_head(vclass.nodes[1]))
       adata = getdata(g[a])
       bdata = getdata(g[b])
-      union!(free, adata)
-      delete!(free, v)
       union!(free, bdata)
-
+      delete!(free, vsy)
+      union!(free, adata)
     elseif op == λ
       v, b = args[1:2]
+      vclass = g[v]
+      vsy = get_constant(g, v_head(vclass.nodes[1]))
       bdata = getdata(g[b])
       union!(free, bdata)
-      delete!(free, v)
-
+      delete!(free, vsy)
     elseif op == Apply
       l, v = args[1:2]
       ldata = getdata(g[l])
       vdata = getdata(g[v])
       union!(free, ldata)
       union!(free, vdata)
-
     end
     return free
   end
 end
 
-EGraphs.join(from::LambdaAnalysis, to::LambdaAnalysis) = union(from, to)
+function EGraphs.join(from::LambdaAnalysis, to::LambdaAnalysis)
+  if issubset(from, to) # includes case from==to
+    from
+  elseif issubset(to, from)
+    to
+  else
+    error("inconsistent free variable sets from: $from to: $to")
+  end
```

Comment:
Is this picking the subset or the superset?

Reply:
Join picks the subset. If there are two expressions in the eclass, one longer and one shorter, the set with fewer free variables is returned.